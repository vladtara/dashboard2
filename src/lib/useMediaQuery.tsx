// Credit: https://www.netlify.com/blog/2020/12/05/building-a-custom-react-media-query-hook-for-more-responsive-apps/
import { useState, useEffect } from "react";

const useMediaQuery = (query: string) => {
  const [matches, setMatches] = useState(false);

  useEffect(() => {
    const media = window.matchMedia(query);
    if (media.matches !== matches) setMatches(media.matches);
    const listener = () => setMatches(media.matches);

    media.addEventListener("change", listener);
    return () => media.removeEventListener("change", listener);
  }, [matches, query]);

  return matches;
};

export const IsDark = () => useMediaQuery("(prefers-color-scheme: dark");

export default useMediaQuery;
