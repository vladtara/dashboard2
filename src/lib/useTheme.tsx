import { IItemProps } from "../components/select";

export interface IThemeProps extends IItemProps {
  mainColor: string;
  accentColor: string;
  backgroundColor: string;
}

export interface IThemeDataProps {
  themes: IThemeProps[];
}

export const defaultTheme: IThemeProps = {
  label: "Classic",
  value: 0,
  mainColor: "#000000",
  accentColor: "#1e272e",
  backgroundColor: "#ffffff",
};

/**
 * Writes the color scheme into localStorage
 * @param {string} scheme
 */
export const setScheme = (scheme: string) => {
  localStorage.setItem("theme", scheme);
};

/**
 * Writes a given theme into localStorage
 * @param {string} scheme - the color scheme (light or dark) to save the theme to
 * @param {string} theme - the theme that shall be saved (in stringified JSON)
 */
export const setTheme = (scheme: string, theme: IThemeProps) => {
  if (scheme === "light") {
    localStorage.setItem("light-theme", JSON.stringify(theme));
    localStorage.setItem("theme", "light-theme");
  }
  if (scheme === "dark") {
    localStorage.setItem("dark-theme", JSON.stringify(theme));
    localStorage.setItem("theme", "dark-theme");
  }
  window.location.reload();
};

/**
 * Function that gets the saved theme from localStorage or returns the default
 * @param {string} [scheme] the color scheme to retrieve the theme for
 * @returns {IThemeProps} the saved theme or the default theme
 */
export const getTheme = (scheme?: string): IThemeProps => {
  let currentScheme = localStorage.getItem("theme");
  let selectedTheme = defaultTheme;

  if (scheme === "light") {
    currentScheme = "light-theme";
  } else if (scheme === "dark") {
    currentScheme = "dark-theme";
  }

  const theme =
    currentScheme === "dark-theme"
      ? localStorage.getItem("dark-theme")
      : localStorage.getItem("light-theme");
  if (theme !== null) selectedTheme = JSON.parse(theme || "{}");

  return selectedTheme;
};

const selectedTheme = getTheme();
export default selectedTheme;
