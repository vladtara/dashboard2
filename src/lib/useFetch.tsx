import { useCallback, useEffect, useState } from "react";

import { IAppListProps } from "../components/appList";
import { IBookmarkListProps } from "../components/bookmarks";
import { ISearchProps } from "../components/searchBar";
import { IThemeDataProps } from "./useTheme";
import { IGreeterDataProps } from "../components/greeter";

export interface IDataProps<I> {
  response?: I;
  error?: string | boolean;
}

const inProduction = process.env.NODE_ENV === "production";

/**
 * Handles the response from the fetch requests
 * @param {Response} response - The response given by the fetch request
 * @returns - The response in JSON
 * @throws - Error with given error message if request failed
 */
export const handleResponse = async (response: Response) => {
  if (!response.ok) throw new Error(response.statusText);
  return response.json();
};

const fetchFile = (f: string) => {
  if (!inProduction) return require(`../data/${f}.json`);

  return fetch(`/data/${f}.json`)
    .then(handleResponse)
    .catch((error: Error) => {
      return { error: error.message };
    });
};

interface IFetchProps {
  appData: IDataProps<IAppListProps>;
  bookmarkData: IDataProps<IBookmarkListProps>;
  searchData: IDataProps<ISearchProps>;
  themeData: IDataProps<IThemeDataProps>;
  greeterData: IDataProps<IGreeterDataProps>;
  callback?: () => void;
}

export const useFetcher = (): IFetchProps => {
  const defaults: IDataProps<any> = { error: true };

  const [appData, setAppData] = useState<IDataProps<IAppListProps>>(defaults);
  const [bookmarkData, setBookmarkData] =
    useState<IDataProps<IBookmarkListProps>>(defaults);
  const [searchData, setSearchData] =
    useState<IDataProps<ISearchProps>>(defaults);
  const [themeData, setThemeData] =
    useState<IDataProps<IThemeDataProps>>(defaults);
  const [greeterData, setGreeterData] =
    useState<IDataProps<IGreeterDataProps>>(defaults);

  const callback = useCallback(() => {
    const files = [
      "apps",
      "bookmarks",
      "search",
      "themes",
      "greeter",
    ];

    Promise.all(files.map((f) => fetchFile(f))).then(
      ([apps, bookmarks, search, themes, greeter]: any) => {
        setAppData({ response: apps });
        setBookmarkData({
          response: bookmarks,
        });
        setSearchData({ response: search });
        setThemeData({ response: themes });
        setGreeterData({ response: greeter });
      },
    );
  }, []);

  useEffect(() => callback(), [callback]);

  return {
    appData,
    bookmarkData,
    searchData,
    themeData,
    greeterData,
  };
};

export default useFetcher;
