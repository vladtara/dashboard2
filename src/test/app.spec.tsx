import { render } from "@testing-library/react";
import App, { GlobalStyle } from "../app";
import { IThemeProps } from "../lib/useTheme";

const props: IThemeProps = {
  label: "Classic",
  value: 0,
  mainColor: "#000000",
  accentColor: "#1e272e",
  backgroundColor: "#ffffff",
};

describe("app.tsx", () => {
  it("Tests GlobalStyle", () => {
    const { baseElement } = render(<GlobalStyle theme={props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
