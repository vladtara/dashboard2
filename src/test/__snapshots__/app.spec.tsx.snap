// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app.tsx Tests GlobalStyle 1`] = `
<body>
  <div />
</body>
`;
